### PR TITLE
dockerTools: split raw and cooked images

### PR DESCRIPTION
**Pull request by tomberek**
Uses passthru and lazy evaluation to separately create raw and cooked
images. Also caches sha256sum of intermediate layers to greatly speed up
creation of larger layers.

###### Motivation for this change
Building large images with multiple layers requires several repeated iterations of tar/untar and sha256sum of layers resulting in O(n^2) performance. This is a WIP to bring much faster builds to dockerTools.

###### Things done
Ideally this should not change existing semantics, just speed up builds and provide some useful caching behavior and passthru attributes.

eg:
```
nix-build -A dockerTools.examples.layersOrder.raw
```

As a WIP there are some extra echo/printf for debugging and understanding purposes.

- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @grahamc @zimbatm @shlevy @adnelson @antoine

### CODE REVIEW

File: pkgs/build-support/docker/default.nix [hunk 1]
```diff
@@ -679,14 +692,38 @@ rec {
                   extraCommands;
         };
       result = runCommand "docker-image-${baseName}.tar.gz" {
+        nativeBuildInputs = [ pigz coreutils ];
+        passthru.buildArgs = args;
+        passthru.layer = layer;
+        passthru.raw = result_raw;
+        passthru.contentsList = let
+          contentsList = if builtins.isList contents then contents else [ contents ];
```
Comment (zimbatm) [inline pkgs/build-support/docker/default.nix]: the lib has a function that does the same thing:
```suggestion
          contentsList = lib.toList contents;
```

File: pkgs/build-support/docker/default.nix [hunk 2]
```diff
@@ -820,10 +870,9 @@ rec {
         # Make the image read-only.
         chmod -R a-w image
 
-        echo "Cooking the image..."
-        tar -C image --hard-dereference --sort=name --mtime="@$SOURCE_DATE_EPOCH" --owner=0 --group=0 --xform s:'^./':: -c . | pigz -nT > $out
-
         echo "Finished."
+        echo "To cook and load:"
+        printf "\e[32m%s\e[0m" 'tar -C result/image --dereference --hard-dereference --sort=name --mtime="1" --owner=0 --group=0 --mode=a-w --xform s:'^./':: -c . | docker load'
```
Comment (zimbatm) [inline pkgs/build-support/docker/default.nix]: More of an idea but this should also work nicely:
```sh
skopeo copy dir:./result/image docker-daemon://image-name:tag-name
```
Skopeo requires to pass the image-name and tag-name to the destination so those could be extracted from the result as well.

**Review by zimbatm — COMMENTED** [review]
On principle, this is great. I always wished that the outputs weren't compressed since it creates so much churn in the nix store and is not really necessary when using skopeo.

This needs a bit more testing to make sure that it works.

File: pkgs/build-support/docker/default.nix (hunk 2 above)
Comment (tomberek) [reply pkgs/build-support/docker/default.nix]: I ran into some problems using skopeo for this purpose related to it expecting Image Manifest V 2, Schema 2. We're using an older schema. But that is/was the goal.

**Comment by tomberek** [comment]
Anything this needs? @grahamc ? Is there a better way to do this, or should this split off into it's own tool (buildRawImage)?

**Comment by zimbatm** [comment]
Yeah it would be easier to merge if it didn't change the existing functions. Or build a set of tests that show that there are no regressions for the old usage.

**Comment by misuzu** [comment]
This is really great! @tomberek any update on this please?

**Comment by purcell** [comment]
I suspect this is superseded by #91084, and could be closed when that is merged?

**Comment by tomberek** [comment]
@purcell, yes. It seems they accomplished a similar goal as well as improving its maintainability.

**Comment by flokli** [comment]
https://github.com/NixOS/nixpkgs/pull/91084 was merged, closing.

**Pull request closed without merging**